src/Model/LanguageFeatures.re

```diff
@@ -0,0 +1,77 @@
+/*
+ * LanguageFeatures.re
+ *
+ * This module documents & types the protocol for communicating with the VSCode Extension Host.
+ */
+
+open Oni_Core;
+
+module CompletionProvider = {
+  type t =
+    (Buffer.t, CompletionMeet.t, Position.t) =>
+    option(Lwt.t(list(CompletionItem.t)));
+};
+
+type t = {completionProviders: list(CompletionProvider.t)};
+
+let empty = {completionProviders: []};
+
+let _identity = v => v;
+
+let requestCompletions =
+    (~buffer: Buffer.t, ~meet: CompletionMeet.t, ~position: Position.t, lf: t) => {
+  let promises =
+    lf.completionProviders
+    |> List.map(suggestor => suggestor(buffer, meet, position))
+    |> Utility.List.filter_map(_identity);
+
+  let completorCount = ref(List.length(promises));
```

Comment:
This is a little messy... I wasn't sure if there was a built-in method in Lwt that does this (I couldn't find it, searching around the docs).

I wanted to return a promise that completes when all promises are resolved, joining the successful results with a function `f` (and in this case, `f` is just list concatenation).

Reply:
Like `Promise.all`? You should be able to do it using `List.fold_left` and `Lwt.bind` I would think. Something like:

```reason
List.fold_left(
  (acc, promise) => {
    let%lwt completions = acc;
    let%lwt completion = promise;
    Lwt.return([completion, ...completions]);
  }),
  Lwt.return([]),
  promises
);
```

FYI: You can drag the "add comment button" to apply a comment to a block of text. It's easier to see what specifically you're referring to then.

Reply:
Cool, that's much nicer, thanks @glennsl !

> FYI: You can drag the "add comment button" to apply a comment to a block of text. It's easier to see what specifically you're referring to then.

Oh.. nice.